### PR TITLE
perf: only check trace in tests if has a reason to

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

Change it to not check `call_tree` unnecessarily (if no `--gas` flag given).
This makes tests run faster when not using `--gas`

**NOTE**: This PR also fixes many race conditions and flakinesses in the tests that came up when adding a test as well as turning off geth snapshotting for now because of this issue: https://github.com/ethereum/go-ethereum/issues/26154

### How I did it

check if needing gas report _before_ checking the grabbing the call tree.

### How to verify it

Can work with Hardhat project with tracing issues and things still work as far as testing goes (obviously sans all tracing features)

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by unparalleled-js** [comment]
Hmm, thinking I may have a better solution, BRB.

**Comment by unparalleled-js** [comment]
- [ ] Fix issue with snapshotting (noticed while testing)

File: src/ape_geth/provider.py [hunk 1]
```diff
@@ -303,8 +304,21 @@ def _stream_request(self, method: str, params: List, iter_path="result.item"):
...
     name: str = "geth"
 
+    @property
+    def chain_id(self) -> int:
+        if self._chain_id is not None:
+            return self._chain_id
+        elif hasattr(self.web3, "eth"):
+            chain_id = self.web3.eth.chain_id
+        else:
+            return GETH_DEV_CHAIN_ID
```
Comment (fubuloubu) [inline src/ape_geth/provider.py:317]: This line won't cache

**Review by helloibis — COMMENTED** [review]
Is this a configuration that should also apply to the dev-message assertion or do we only mean tracing here as in gas tracing / profiling?

**Comment by unparalleled-js** [comment]
> Is this a configuration that should also apply to the dev-message assertion or do we only mean tracing here as in gas tracing / profiling?

I removed the config because I found out it wasn't needed! So no worries there :)

File: src/ape_geth/provider.py [hunk 2]
```diff
@@ -349,18 +367,41 @@ def disconnect(self):
 
         super().disconnect()
 
+    @raises_not_implemented
+    def snapshot(self) -> SnapshotID:
+        # TODO: Replace with impl below after
+        #  https://github.com/ethereum/go-ethereum/issues/26154 resolved
```
Comment (NotPeopling2day) [inline src/ape_geth/provider.py:373]: Do we have a ticket to track this?

**Review by NotPeopling2day — APPROVED** [review]

**Review by helloibis — APPROVED** [review]

File: src/ape_geth/provider.py (hunk 2 above)
Comment (unparalleled-js) [reply src/ape_geth/provider.py:373]: APE-319

**Pull request merged**